linera-views/src/register_view.rs

```diff
@@ -58,29 +62,45 @@ where
     }
 
     fn rollback(&mut self) {
+        self.was_cleared = false;
+        self.after_clear_set = false;
         self.update = None;
         *self.hash.get_mut() = self.stored_hash;
     }
 
     fn flush(&mut self, batch: &mut Batch) -> Result<(), ViewError> {
-        if let Some(value) = self.update.take() {
-            let key = self.context.base_tag(KeyTag::Value as u8);
-            batch.put_key_value(key, &value)?;
-            self.stored_value = value;
+        if self.was_cleared {
+            batch.delete_key_prefix(self.context.base_key());
+        }
+        // If there is clearing then we want to write the value only if it has been set afterward
+        if !self.was_cleared || self.after_clear_set {
+            if let Some(value) = self.update.take() {
+                let key = self.context.base_tag(KeyTag::Value as u8);
+                batch.put_key_value(key, &value)?;
+                self.stored_value = value;
+            }
         }
         let hash = *self.hash.get_mut();
-        if self.stored_hash != hash {
+        // In the scenario where we do a clear
+        // and stored_hash = hash, we need to update the
+        // hash, otherwise, we will recompute it while this
+        // can be avoided.
+        if self.stored_hash != hash || self.was_cleared {
             let key = self.context.base_tag(KeyTag::Hash as u8);
             match hash {
                 None => batch.delete_key(key),
                 Some(hash) => batch.put_key_value(key, &hash)?,
             }
             self.stored_hash = hash;
         }
+        self.was_cleared = false;
+        self.after_clear_set = false;
         Ok(())
     }
 
     fn clear(&mut self) {
+        self.was_cleared = true;
+        self.after_clear_set = false;
         self.update = Some(Box::default());
```

Comment:
Also, we could have `self.update = None` here.

Reply:
Yes, the `self.update = None` simplifies the code.

Reply:
I was wrong about `self.update = None` (see new comments)

Reply:
Ok, reverted.